Eval: Seating Arrangement

floridaman opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, __failure to follow the guidelines below will result in the PR being closed automatically__. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access granted. 🚨

__PLEASE READ THIS__:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, pelase note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑
### Eval name
Seating Arrangements

### Eval description

An array of seating arrangement constraints, each with 2 solutions. Examining the models spatial reasoning abilities.

### What makes this a useful eval?

Assists with evaluating simple spatial reasoning with a limited number of solutions to a constraint problem. I created a solver to ensure that each puzzle only has 2 solutions and a limited number of variables. A human should be able to solve these extremely easily even without writing the names down on pen and paper.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [x] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can  create an eval on cases where the model fails to reason about the physical world.
- [x] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [x] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [x] **Include at least 15 high quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

> Insert what makes your eval high quality that was not mentioned above. (Not required)

## Eval structure 🏗️

Your eval should
- [x] Check that your data is in `evals/registry/data/{name}`
- [x] Check that your yaml is registered at `evals/registry/evals/{name}.yaml`
- [x] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (https://platform.openai.com/docs/usage-policies).

- [x] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the merged pull request.

- [x] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgement

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [x] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access granted.

### Submit eval

- [x] I have filled out all required fields of this form
- [x] I have used **Git LFS** for the Eval JSON data
- [x] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data 

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
  {"input": [{"role": "system", "content": "You are a helpful assistant who creates seating arrangements for people in a single row at a long table. Loners want to sit at one end of the table, friends need to be seated next to each other, and enemies should not be seated next to each other. You will print seating arrangements in the form of person names separated by the | character like so Tim | Jeremey | Ana."}, {"role": "user", "content": "Please provide a valid seating arragement for the following requirements: Daniel and Sophia are enemies, Sophia and Charlotte are enemies, Olivia and Daniel are friends, Daniel and Benjamin are friends, Benjamin and Charlotte are friends."}], "ideal": ["Charlotte | Benjamin | Daniel | Olivia | Sophia", "Sophia | Olivia | Daniel | Benjamin | Charlotte"]}
{"input": [{"role": "system", "content": "You are a helpful assistant who creates seating arrangements for people in a single row at a long table. Loners want to sit at one end of the table, friends need to be seated next to each other, and enemies should not be seated next to each other. You will print seating arrangements in the form of person names separated by the | character like so Tim | Jeremey | Ana."}, {"role": "user", "content": "Please provide a valid seating arragement for the following requirements: Eleanor is a loner, Samuel and Christopher are enemies, Emma is a loner, Samuel and Eleanor are enemies."}], "ideal": ["Eleanor | Christopher | Scarlett | Samuel | Emma", "Emma | Samuel | Scarlett | Christopher | Eleanor"]}
{"input": [{"role": "system", "content": "You are a helpful assistant who creates seating arrangements for people in a single row at a long table. Loners want to sit at one end of the table, friends need to be seated next to each other, and enemies should not be seated next to each other. You will print seating arrangements in the form of person names separated by the | character like so Tim | Jeremey | Ana."}, {"role": "user", "content": "Please provide a valid seating arragement for the following requirements: Nathan and Samuel are enemies, David and Samuel are friends, Olivia and Samuel are friends, Olivia is a loner, David and Victoria are friends."}], "ideal": ["Nathan | Victoria | David | Samuel | Olivia", "Olivia | Samuel | David | Victoria | Nathan"]}
{"input": [{"role": "system", "content": "You are a helpful assistant who creates seating arrangements for people in a single row at a long table. Loners want to sit at one end of the table, friends need to be seated next to each other, and enemies should not be seated next to each other. You will print seating arrangements in the form of person names separated by the | character like so Tim | Jeremey | Ana."}, {"role": "user", "content": "Please provide a valid seating arragement for the following requirements: Olivia and Eleanor are enemies, Grace and Olivia are friends, Eleanor and Alexander are enemies, Scarlett and Olivia are friends, Grace and Eleanor are enemies."}], "ideal": ["Eleanor | Scarlett | Olivia | Grace | Alexander", "Alexander | Grace | Olivia | Scarlett | Eleanor"]}
{"input": [{"role": "system", "content": "You are a helpful assistant who creates seating arrangements for people in a single row at a long table. Loners want to sit at one end of the table, friends need to be seated next to each other, and enemies should not be seated next to each other. You will print seating arrangements in the form of person names separated by the | character like so Tim | Jeremey | Ana."}, {"role": "user", "content": "Please provide a valid seating arragement for the following requirements: Ava and Christopher are friends, Isaac and Benjamin are enemies, David and Ava are friends, Christopher and Benjamin are enemies."}], "ideal": ["Benjamin | David | Ava | Christopher | Isaac", "Isaac | Christopher | Ava | David | Benjamin"]}
  ```
</details>

usama-openai reviewed (changes requested):
Thanks for the contribution. This PR looks good. However, I would like to request some changes.

1. The evaluation method being used to evaluate the answer is `Match` which only works when an exact match is found. The user needs to provide very specific instructions to the model about the output format, and the ideal answer should be in that expected format. In this dataset, the model is being instructed to provide seating arrangements with names separated by "|". There are no specific instructions about providing only the arrangement and nothing else, and the model may provide an answer with some explanation or reasoning. In such scenarios, `Match` would fail despite the right answer being provided by the model.

    You need to either provide proper prompt instructions about the output, and the ideal answer should be formatted according to the prompt instructions, or you should use `Includes` here, which checks if the ideal answer is included in the model completion.

2. Add a proper description of the eval in the `.yaml` file.

We would love to review the PR again after the suggested changes.

floridaman commented:
@usama-openai I modified the eval template to utilize the Includes template and verified the output. I have also added a description to the .yaml.

In the future is it preferable to have more precise prompting for the output or to utilize Includes to ensure the desired solution is at least somewhere in the output?

Let me know if you find anything else.

usama-openai commented:
It depends on the nature of the task. For simple tasks, models are able to solve them without any reasoning, and `Match` can work fine. Some tasks are complex and involve multistep reasoning. For such tasks, the model should be allowed to reason before providing the answer. This gives the model a fair chance to solve it, and `Includes` or `FuzzyMatch` can be used to compare ideal answers.

usama-openai reviewed (changes requested):
Thanks for the updates. It appears that in the previous review, I forgot to ask for one change.

The naming convention for evals is in the form `<eval_name>.<split>.<version>`.
- `<eval_name>` is the eval name, used to group evals whose scores are comparable.
- `<split>` is the data split, used to further group evals that are under the same <base_eval>. E.g., val, test, or dev for testing.
- `<version>` is the version of the eval, which can be any descriptive text you'd like to use (though it's best if it does not contain.).

You can rename the eval id `seating_arrangements.v0` to `seating_arrangements.dev.v0`.

floridaman commented:
@usama-openai I modified the name of the eval.

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.
